solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LambdaMARTModel.java

```diff
@@ -86,123 +140,138 @@ public String explain(float[] featureVector) {
-        // that use a feature that is no longer output during feature extraction
-        // TODO: make lambdaSmart (in rank_svm_final repo )
-        // either remove trees that depend on such features
-        // or prune them back above the split on that feature
-        featureIndex = (idx == null) ? -1 : idx;
-
-        final Object ot = map.get("threshold");
-        if (null == ot) {
-          throw new ModelException(
-              "LambdaMARTModel tree node is missing threshold");
+        sb.append("(feature=").append(feature);
+        if (threshold != null) {
+          sb.append(",threshold=").append(threshold.floatValue()-NODE_SPLIT_SLACK);
+        } else {
+          sb.append(",threshold=").append(threshold);
         }
+        sb.append(",left=").append(left);
+        sb.append(",right=").append(right);
+        sb.append(')');
+      }
+      return sb.toString();
+    }
 
-        threshold = LTRUtils.convertToFloat(ot) + NODE_SPLIT_SLACK;
+    public RegressionTreeNode() {
+    }
 
-        final Object ol = map.get("left");
-        if (null == ol) {
-          throw new ModelException("LambdaMARTModel tree node is missing left");
+    public void validate() throws ModelException {
+      if (isLeaf()) {
+        if (left != null || right != null) {
+          throw new ModelException("LambdaMARTModel tree node is leaf with left="+left+" and right="+right);
         }
+        return;
+      }
+      if (null == threshold) {
+        throw new ModelException("LambdaMARTModel tree node is missing threshold");
+      }
+      if (null == left) {
+        throw new ModelException("LambdaMARTModel tree node is missing left");
+      } else {
+        left.validate();
+      }
+      if (null == right) {
+        throw new ModelException("LambdaMARTModel tree node is missing right");
+      } else {
+        right.validate();
+      }
+    }
 
-        left = new RegressionTreeNode((Map<String,Object>) ol, fname2index);
+  }
 
-        final Object or = map.get("right");
-        if (null == or) {
-          throw new ModelException("LambdaMARTModel tree node is missing right");
-        }
+  public class RegressionTree {
 
-        right = new RegressionTreeNode((Map<String,Object>) or, fname2index);
-      }
+    private Float weight;
+    private RegressionTreeNode root;
+
+    public void setWeight(float weight) {
+      this.weight = new Float(weight);
     }
 
-  }
+    public void setWeight(String weight) {
+      this.weight = new Float(weight);
+    }
 
-  class RegressionTree {
-    private final float weight;
-    private final RegressionTreeNode root;
+    public void setTree(Object root) {
+      this.root = createRegressionTreeNode((Map<String,Object>)root);
+    }
 
     public float score(float[] featureVector) {
-      return weight * root.score(featureVector);
+      return (weight == null || root == null ? 0f : weight.floatValue() * root.score(featureVector));
```

Comment:
remove all the null checks (and make sure validate throws an exception if it happens)